### PR TITLE
[Espresso] Re-compute StaticPropertyKind.Object offsets for SVM

### PR DESCRIPTION
**Pull request by ivan-ristovic**
This PR is a pre-requisite for `LinkedKlass` caching during context pre-initialization (see https://github.com/oracle/graal/pull/3482). These offsets need to be re-computed so cached `LinkedKlass`es can be re-used on SVM.

### CODE REVIEW

**Comment by mukel** [comment]
From @ansalond :
The changes look good. Some feedback:
-    move to the static init as much computation as possible and store results in static fields. In particular, calls to Class.forName(), method lookups, and the call to StaticPropertyKind.Object.toByte() . It's rather ugly to have static fields, but native image is often blamed for being slow and I know C. Wimmer is very careful about these things.
-    } catch (ReflectiveOperationException e) { (for consistency with the rest)
- I'm not sure we can import sun.misc.Unsafe . It looks reasonable here, but I don't know if it is vetoed

There's a test that only check primitive fields, it can be updated as follows:
```
diff --git a/truffle/src/com.oracle.truffle.api.staticobject.test/src/com/oracle/truffle/api/staticobject/test/StaticTest.java b/truffle/src/com.oracle.truffle.api.staticobject.test/src/com/oracle/truffle/api/staticobject/test/StaticTest.java
index 98ee4aa65c5e..3c3aac2aade7 100644
--- a/truffle/src/com.oracle.truffle.api.staticobject.test/src/com/oracle/truffle/api/staticobject/test/StaticTest.java
+++ b/truffle/src/com.oracle.truffle.api.staticobject.test/src/com/oracle/truffle/api/staticobject/test/StaticTest.java
@@ -53,22 +53,32 @@ import org.junit.Test;
  * built time for context pre-initialization.
  */
 public class StaticTest {
-    private static final StaticProperty property;
+    private static final StaticProperty property1;
+    private static final StaticProperty property2;
     private static final Object staticObject;
+    private static final Object testValue1 = "object1";
+    private static final Object testValue2 = "object2";
 
     static {
         TestEnvironment environment = new TestEnvironment(new TestConfiguration(true, false));
         StaticShape.Builder builder = StaticShape.newBuilder(environment.testLanguage);
-        property = new DefaultStaticProperty("property");
-        builder.property(property, int.class, false);
+        property1 = new DefaultStaticProperty("property1");
+        property2 = new DefaultStaticProperty("property2");
+        builder.property(property1, Object.class, false);
+        builder.property(property2, Object.class, false);
         staticObject = builder.build().getFactory().create();
-        property.setInt(staticObject, 42);
+        property1.setObject(staticObject, testValue1);
+        property2.setObject(staticObject, testValue2);
     }
 
     @Test
     public void staticallyDeclaredStaticObject() {
-        Assert.assertEquals(42, property.getInt(staticObject));
-        property.setInt(staticObject, 24);
-        Assert.assertEquals(24, property.getInt(staticObject));
+        Assert.assertEquals(testValue1, property1.getObject(staticObject));
+        Assert.assertEquals(testValue2, property2.getObject(staticObject));
+        Object newTestValue = "object3";
+        property1.setObject(staticObject, newTestValue);
+        Assert.assertEquals(newTestValue, property1.getObject(staticObject));
+        property2.setObject(staticObject, newTestValue);
+        Assert.assertEquals(newTestValue, property2.getObject(staticObject));
     }
 }
```

**Comment by ivan-ristovic** [comment]
Thanks for the quick review!

I have added the requested changes. As for `Unsafe`, I left it as-is for now, until someone else takes a look.

File: substratevm/src/com.oracle.svm.truffle/src/com/oracle/svm/truffle/TruffleBaseFeature.java [hunk 1]
```diff
@@ -625,6 +628,62 @@ private static void registerReflectionAccessesForRuntimeValidation(Class<?> stor
...
+
+        @Override
+        public Object transform(MetaAccessProvider metaAccess, ResolvedJavaField original, ResolvedJavaField annotated,
+                        Object receiver, Object originalValue) {
+            int offset = (int) originalValue;
+            StaticProperty receiverStaticProperty = (StaticProperty) receiver;
+            try {
+                byte internalKindByte = (byte) staticPropertyKindGetInternalKind.invoke(receiverStaticProperty);
+
+                // Re-computation is performed only if the kind is Object:
```
Comment (vjovanov) [inline substratevm/src/com.oracle.svm.truffle/src/com/oracle/svm/truffle/TruffleBaseFeature.java:669]: We use `/**/` style comments.

File: substratevm/src/com.oracle.svm.truffle/src/com/oracle/svm/truffle/TruffleBaseFeature.java [hunk 2]
```diff
@@ -625,6 +628,62 @@ private static void registerReflectionAccessesForRuntimeValidation(Class<?> stor
...
 
+@TargetClass(className = "com.oracle.truffle.api.staticobject.StaticProperty", onlyWith = TruffleBaseFeature.IsEnabled.class)
+final class Target_com_oracle_truffle_api_staticobject_StaticProperty {
+    public static final class OffsetTransformer implements RecomputeFieldValue.CustomFieldValueTransformer {
+
+        private static final int svmArrayBaseOffset;
+        private static final int svmArrayIndexScale;
+
+        // We have to use reflection to access private members instead of aliasing them
+        // in the substitution class since substitutions are present only in runtime
```
Comment (vjovanov) [inline substratevm/src/com.oracle.svm.truffle/src/com/oracle/svm/truffle/TruffleBaseFeature.java:639]: at runtime.

File: substratevm/src/com.oracle.svm.truffle/src/com/oracle/svm/truffle/TruffleBaseFeature.java [hunk 3]
```diff
@@ -625,6 +628,62 @@ private static void registerReflectionAccessesForRuntimeValidation(Class<?> stor
...
+        private static final Method staticPropertyKindGetInternalKind;
+        private static final byte objectStaticPropertyKindByte;
+
+        static {
```
Comment (vjovanov) [inline substratevm/src/com.oracle.svm.truffle/src/com/oracle/svm/truffle/TruffleBaseFeature.java:643]: This is ugly, but I guess the "right" way is to extract the API to the Graal SDK and use it in the `com.oracle.truffle.api`? That would be out of the scope of this PR.

**Review by vjovanov — APPROVED** [review]
LGTM

File: substratevm/src/com.oracle.svm.truffle/src/com/oracle/svm/truffle/TruffleBaseFeature.java (hunk 1 above)
Comment (ivan-ristovic) [reply substratevm/src/com.oracle.svm.truffle/src/com/oracle/svm/truffle/TruffleBaseFeature.java]: Changed. I also felt free to update the rest of the comments in this file that did not follow this convention.

File: substratevm/src/com.oracle.svm.truffle/src/com/oracle/svm/truffle/TruffleBaseFeature.java (hunk 3 above)
Comment (ivan-ristovic) [reply substratevm/src/com.oracle.svm.truffle/src/com/oracle/svm/truffle/TruffleBaseFeature.java:646]: As we agreed, one reflection call to an internal method is fine, so I have re-structured the code accordingly.

**Comment by ivan-ristovic** [comment]
@mukel Could you please re-review and update the internal PR state? Thanks!

File: substratevm/src/com.oracle.svm.truffle/src/com/oracle/svm/truffle/TruffleBaseFeature.java [hunk 4]
```diff
@@ -625,6 +630,112 @@ private static void registerReflectionAccessesForRuntimeValidation(Class<?> stor
...
+                    indexScale = Unsafe.ARRAY_SHORT_INDEX_SCALE;
+                    break;
+                default:
+                    return originalValue;
+            }
+
+            /*
+             * Reverse the offset computation to find the index
+             */
+            int index = (offset - baseOffset) / indexScale;
```
[the remark this answers is not in the document]
Comment (ivan-ristovic) [reply substratevm/src/com.oracle.svm.truffle/src/com/oracle/svm/truffle/TruffleBaseFeature.java:685]: I added a simple assertion which will check this.

File: truffle/src/com.oracle.truffle.api.staticobject.test/src/com/oracle/truffle/api/staticobject/test/StaticTest.java [hunk 5]
```diff
@@ -53,22 +53,32 @@
...
+    private static final StaticProperty property2;
     private static final Object staticObject;
+    private static final Object testValue1 = "object1";
+    private static final Object testValue2 = "object2";
 
     static {
         TestEnvironment environment = new TestEnvironment(new TestConfiguration(true, false));
         StaticShape.Builder builder = StaticShape.newBuilder(environment.testLanguage);
-        property = new DefaultStaticProperty("property");
-        builder.property(property, int.class, false);
```
[the remark this answers is not in the document]
Comment (ivan-ristovic) [reply truffle/src/com.oracle.truffle.api.staticobject.test/src/com/oracle/truffle/api/staticobject/test/StaticTest.java:63]: Added tests for each primitive type.

File: substratevm/src/com.oracle.svm.truffle/src/com/oracle/svm/truffle/TruffleBaseFeature.java [hunk 6]
```diff
@@ -625,6 +630,112 @@ private static void registerReflectionAccessesForRuntimeValidation(Class<?> stor
...
+                internalKindName = (String) staticPropertyGetInternalKind.invoke(receiverStaticProperty);
+                // Checkstyle: resume
+            } catch (IllegalAccessException | InvocationTargetException e) {
+                throw VMError.shouldNotReachHere(e);
+            }
+
+            int baseOffset;
+            int indexScale;
+            JavaKind javaKind;
+            switch (internalKindName) {
```
[the remark this answers is not in the document]
Comment (ivan-ristovic) [reply substratevm/src/com.oracle.svm.truffle/src/com/oracle/svm/truffle/TruffleBaseFeature.java]: Yes, a simple `if` statement comparing the `internalKindName` with `Object` and using `Object` offsets in such case and `byte` offsets in all other cases will also work fine. I updated the re-computation based on the assumption that SVM will use the same base offsets for all primitives. The only reason why initially I used a switch statement was maintenance in case this ever changes.

File: substratevm/src/com.oracle.svm.truffle/src/com/oracle/svm/truffle/TruffleBaseFeature.java [hunk 7]
```diff
@@ -735,22 +741,53 @@ public Object transform(MetaAccessProvider metaAccess, ResolvedJavaField origina
...
 
-// If allowProcess() is disabled at build time, then we ensure that
-// ProcessBuilder is not reachable.
-// The main purpose of this is to test that ProcessBuilder is not part of the
-// image when building
-// language images with allowProcess() disabled, which we interpret as "forbid
-// shelling out to
-// external processes" (GR-14041).
+@TargetClass(className = "com.oracle.truffle.polyglot.PolyglotContextImpl", onlyWith = TruffleBaseFeature.IsEnabled.class)
+final class Target_com_oracle_truffle_polyglot_PolyglotContextImpl {
```
Comment (gilles-duboscq) [inline substratevm/src/com.oracle.svm.truffle/src/com/oracle/svm/truffle/TruffleBaseFeature.java:753]: That's new, is it really for this PR? It should at least be in its own commit

File: substratevm/src/com.oracle.svm.truffle/src/com/oracle/svm/truffle/TruffleBaseFeature.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (gilles-duboscq) [reply substratevm/src/com.oracle.svm.truffle/src/com/oracle/svm/truffle/TruffleBaseFeature.java]: There's no assumption there: the new code is correct, _especially_ if SVM starts using different base offsets: primitives are all stored in a byte array so we must use the offset and scales of byte arrays for both HotSpot and SVM.

File: substratevm/src/com.oracle.svm.truffle/src/com/oracle/svm/truffle/TruffleBaseFeature.java (hunk 7 above)
Comment (ivan-ristovic) [reply substratevm/src/com.oracle.svm.truffle/src/com/oracle/svm/truffle/TruffleBaseFeature.java]: This is a mistake introduced by the rebase (there were conflicts with the master). Should be resolved now.

**Review by gilles-duboscq — APPROVED** [review]

**Review by ansalond — CHANGES_REQUESTED** [review]

File: substratevm/src/com.oracle.svm.truffle/src/com/oracle/svm/truffle/TruffleBaseFeature.java [hunk 8]
```diff
@@ -625,6 +627,74 @@ private static void registerReflectionAccessesForRuntimeValidation(Class<?> stor
...
+        static {
+            // Checkstyle: stop
+            staticPropertyGetInternalKind = ReflectionUtil.lookupMethod(StaticProperty.class, "getInternalKindName");
+            // Checkstyle: resume
+        }
+
+        @Override
+        public Object transform(MetaAccessProvider metaAccess, ResolvedJavaField original, ResolvedJavaField annotated,
+                        Object receiver, Object originalValue) {
+            int offset = (int) originalValue;
```
[the remark this answers is not in the document]
Comment (ivan-ristovic) [reply substratevm/src/com.oracle.svm.truffle/src/com/oracle/svm/truffle/TruffleBaseFeature.java:652]: Added a check! Thanks!

**Review by ansalond — APPROVED** [review]
Thanks for the contribution! I'm integrating this PR.
